sys/net/gnrc/transport_layer/tcp/gnrc_tcp_pkt.c

```diff
@@ -0,0 +1,435 @@
+
+/* Check if a sequence number, falls into the receive window */
+#define INSIDE_WND(l_ed, seq_num, r_ed) (LEQ_32_BIT(l_ed, seq_num) && LSS_32_BIT(seq_num, r_ed))
+
+/**
+ * @brief Calculates the maximum of two unsigned numbers
+ *
+ * @param [in] x   First comparrison value
+ * @param [in] y   Second comparrison value
+ *
+ * @return   x if x is larger than y, if not y is returned.
+ */
+static inline uint32_t _max(const uint32_t x, const uint32_t y)
+{
+  return (x > y) ? x : y;
+}
+
+int8_t _pkt_build_reset_from_pkt(gnrc_pktsnip_t **out_pkt, gnrc_pktsnip_t *in_pkt)
+{
+    gnrc_pktsnip_t *tcp_snp = NULL;
+    gnrc_pktsnip_t *ip6_snp = NULL;
+    tcp_hdr_t tcp_hdr_out;
+
+    /* Extract headers */
+    LL_SEARCH_SCALAR(in_pkt, tcp_snp, type, GNRC_NETTYPE_TCP);
+    tcp_hdr_t *tcp_hdr_in = (tcp_hdr_t *)tcp_snp->data;
+#ifdef MODULE_GNRC_IPV6
+    LL_SEARCH_SCALAR(in_pkt, ip6_snp, type, GNRC_NETTYPE_IPV6);
+    ipv6_hdr_t *ip6_hdr = (ipv6_hdr_t *)ip6_snp->data;
+#endif
+
+    /* Setup Header information */
+    tcp_hdr_out.src_port = tcp_hdr_in->dst_port;
+    tcp_hdr_out.dst_port = tcp_hdr_in->src_port;
+    tcp_hdr_out.checksum = byteorder_htons(0);
+    tcp_hdr_out.window = byteorder_htons(0);
+    tcp_hdr_out.urgent_ptr = byteorder_htons(0);
+
+    /* Seq/Ackno and control flags depend on inputs ACK-Flag */
+    uint16_t ctl = byteorder_ntohs(tcp_hdr_in->off_ctl);
+    if (ctl & MSK_ACK) {
+        tcp_hdr_out.off_ctl = byteorder_htons((OPTION_OFFSET_BASE << 12) | MSK_RST);
+        tcp_hdr_out.seq_num = tcp_hdr_in->ack_num;
+        tcp_hdr_out.ack_num = byteorder_htonl(0);
+    }
+    else {
+        uint8_t seq_no = 0;
+        tcp_hdr_out.off_ctl = byteorder_htons((OPTION_OFFSET_BASE << 12) | MSK_RST_ACK);
+        tcp_hdr_out.seq_num = byteorder_htonl(0);
+        if (ctl & MSK_SYN) {
+            seq_no += 1;
+        }
+        if (ctl & MSK_FIN) {
+            seq_no += 1;
+        }
+        uint32_t tmp = byteorder_ntohl(tcp_hdr_in->seq_num);
+        tcp_hdr_out.ack_num = byteorder_htonl(seq_no + tmp  + _pkt_get_pay_len(in_pkt));
+    }
+
+    /* Allocate new tcp header */
+    tcp_snp = gnrc_pktbuf_add(NULL, &tcp_hdr_out, OPTION_OFFSET_BASE * 4, GNRC_NETTYPE_TCP);
+    if (tcp_snp == NULL) {
+        DEBUG("gnrc_tcp_pkt : _pkt_build_reset_from_pkt() : Can't alloc buffer for TCP Header\n.");
+        *(out_pkt) = NULL;
+        return -ENOMEM;
+    }
+    else {
+        *out_pkt = tcp_snp;
+    }
+
+    /* Build new network layer header */
+#ifdef MODULE_GNRC_IPV6
+    ip6_snp = gnrc_ipv6_hdr_build(tcp_snp, &(ip6_hdr->dst), &(ip6_hdr->src));
+    if (ip6_snp == NULL) {
+        DEBUG("gnrc_tcp_pkt : _pkt_build_reset_from_pkt() : Can't alloc buffer for IPv6 Header.\n");
+        gnrc_pktbuf_release(tcp_snp);
+        *(out_pkt) = NULL;
+        return -ENOMEM;
+    }
+    else {
+        *out_pkt = ip6_snp;
+    }
+#endif
+    return 0;
+}
+
+int8_t _pkt_build(gnrc_tcp_tcb_t* tcb, gnrc_pktsnip_t **out_pkt, uint16_t *seq_con,
```

Comment:
not exactly, because it states:

> when you need to make sure that a certain variable is exactly of this length

I don't think this applies here, its only return value and in other places we use a simple `int` that suffice here too.